paddle/fluid/operators/center_loss_op.cc

```diff
@@ -0,0 +1,152 @@
+
+    PADDLE_ENFORCE(ctx->HasInput("Label"),
+                   "Input(Label) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasInput("Centers"),
+                   "Input(Centers) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("CentersDiff"),
+                   "Output(CentersDiff) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(ctx->HasOutput("Loss"),
+                   "Output(Loss) of CenterLoss should not be null.");
+
+    PADDLE_ENFORCE(
+        ctx->HasOutput("CentersOut"),
+        "Output(CentersOut) of CenterLoss shared data with Centers.");
+
+    ctx->SetOutputDim("CentersDiff", {x_dims[0], product(x_dims) / x_dims[0]});
+    ctx->SetOutputDim("CentersOut", ctx->GetInputDim("Centers"));
+    ctx->SetOutputDim("Loss", {x_dims[0], 1});
+    ctx->ShareLoD("X", /*->*/ "Loss");
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext &ctx) const override {
+    return framework::OpKernelType(ctx.Input<Tensor>("X")->type(),
+                                   ctx.device_context());
+  }
+};
+
+class CenterLossOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("Label", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("Centers", "(Tensor) Input tensor of center_loss operator.");
+    AddInput("CenterUpdateRate",
+             "(Tensor) Input tensor of center_loss operator.");
+
+    AddOutput("CentersOut", "(Tensor) Input tensor of center_loss operator.");
+    AddOutput("CentersDiff", "(Tensor) output tensor of center_loss operator.");
+    AddOutput("Loss", "(Tensor) Output tensor of center_loss operator.");
+
+    AddAttr<int>("cluster_num",
+                 "The output cluster num of the center_loss operator.");
+    AddAttr<bool>("need_update", "whether need to update center info.");
+    AddComment(R"DOC(
+**CenterLoss operator**
+implemention of the algorithm in the papper A Discriminative 
+Feature Learning Approach for Deep Face Recognition
+)DOC");
+  }
+};
+
+class CenterLossGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext *ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("CentersDiff"),
+                   "Input(CentersDiff) should not be null");
+    PADDLE_ENFORCE(ctx->HasInput(framework::GradVarName("Loss")),
+                   "Input(Loss) should not be null");
+    PADDLE_ENFORCE(ctx->HasOutput(framework::GradVarName("X")),
+                   "Output(X) should not be null");
+
+    auto x_dims = ctx->GetInputDim("X");
+    auto x_grad_name = framework::GradVarName("X");
+
+    if (ctx->HasOutput(x_grad_name)) {
+      ctx->SetOutputDim(x_grad_name, x_dims);
+    }
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
```

Comment:
As the type is consistent among all the inputs and outputs, so it not required to specify a `GetExpectedKernelType` methods. But it is OK to keep.